### DISCUSSION TITLE
Init of the Couchbasev4 extension

### DISCUSSION THREAD

**Discussion by Geolim4** [TruCopilot/couchbasev4-extension#1] [General]
Follow up of the initial [pull request](https://github.com/PHPSocialNetwork/phpfastcache/pull/905) on main Phpfastcache's repository.
Ping @srjlewis srjlewis

**Comment by Geolim4** [comment]
Please also make a check of the PHP extension to make sure that at least the 4.0 is used for this driver.

**Reply by Geolim4** [reply]
Damn, I missed the part of the code you made for this:
https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/aed42f08455de6261658299d25663b881e526895/lib/Phpfastcache/Drivers/Couchbasev4/Driver.php#L70

You can simplify it using `str_starts_with` on '4.' string :P

**Reply by srjlewis** [reply]
lol I could of done be went all the way

**Comment by srjlewis** [comment]
I am going to get this pulled into my framework, put it to work and add some of my latest changes.

Main change is to the Config class, I have removed the proxing to the ClusterOptions, it looked like a good idea at the time, but over complex for documentaion, maintaince etc.

I will also added most of the changes from 9.2

**Comment by Geolim4** [comment]
I'm gonna push some changes to the main repo for the extension mechanism.

**Comment by Geolim4** [comment]
https://github.com/PHPSocialNetwork/phpfastcache/commit/e0223f0822f514d4f1df5867333b664b4c147c3f

**Comment by Geolim4** [comment]
If things work properly:
https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/8f854ad52f6defbf8db366c8aafe21930bb1280b/lib/ExtensionRegister.php#L7

The file is auto-loaded automatically and will register the extension to the `ExtensionManager.`

Then the CacheManager will be able to automatically discover and load the extension:
https://github.com/PHPSocialNetwork/phpfastcache/blob/e0223f0822f514d4f1df5867333b664b4c147c3f/lib/Phpfastcache/CacheManager.php#L119

**Reply by srjlewis** [reply]
I was just reading through the commits, I will be doing a intergration with it tomorrow to see how every thing goes, so fingers crossed it will go smoothly, I do like the extention model concept has it removes a load of requires from composer, which are not needed if you are not useing those drivers.

**Comment by Geolim4** [comment]
That's why I put a lot of dependencies as "suggest" until now (the require list push was an error).

**Reply by srjlewis** [reply]
oppps, I pretend I did not see it 🫣

**Comment by Geolim4** [comment]
For more consitency, extension version will have the same version than Phpfastcache.

So technically `phpfastcache/couchbasev4-extension` will starts at **v9.2.0-alpha** (which I just pushed the tag).

**Reply by srjlewis** [reply]
Yes that makes sense to begin with, but going forward are we going to retag each extention on each release of the core, even if there has been no changes to the driver/extention, and what all we need to do is a update to a extention with out the need for a release of the core.

**Reply by Geolim4** [reply]
I don't plan to go that far. Only to restrict extensions to each MAJOR versions of phpfastcache.
A 9.2.1 version of Phpfastcacge will not necessarily trigger a 9.2.1 of `phpfastcache/couchbasev4-extension` as Phpfastcache lib is SEMVER compliant. However a retag will be necessary if the major version of Phpfastcache is released :)

`phpfastcache/couchbasev4-extension` will follow its own versions, but the major version will follow the Phpfastcache core.

**Reply by Geolim4** [reply]
That's why I released an API Changelog years ago:

https://github.com/PHPSocialNetwork/phpfastcache/blob/master/CHANGELOG_API.md

To display public interfaces changes that could breaks compatibility between each MAJOR versions.

**Reply by srjlewis** [reply]
Yes total agree, that makes much more sense

**Comment by Geolim4** [comment]
For now I don't know how to make the test suite working for this extension, I need to see if I can reuse the tests lib from Phpfastcache and if I can make a Gitlab Action test suite with Couchbasev4

**Reply by Geolim4** [reply]
I added `webmozart/assert` as a dev dep so you can make simple tests for Github Actions.

**Comment by Geolim4** [comment]
I finally took time to setup the Github Action 🥳 
https://github.com/PHPSocialNetwork/couchbasev4-extension/actions/runs/7428265248/job/20215339882

**Reply by Geolim4** [reply]
I just pushed a commit to make those extensions requirements optional but required with the `doPosixCheck` config:
https://github.com/PHPSocialNetwork/couchbasev4-extension/commit/fd8f2fef1d874372c8bc30c0dba53aed60cbee14

**Reply by Geolim4** [reply]
It's late night but I finally managed to split up my test tools into a separate repository so I can use it both in core and extensions repos :D
Gotta go sleep now !

**Reply by srjlewis** [reply]
Morning, sorry I was not about last night (I had family stuff to do)

I did what to use the existing test tools, but I could not pull them in, So I am glad you have split the test tools out and removed the messy test code :)

You beat me the the `setDoPosixCheck`, I was going to use something on the lines of `setForkDetection` has it would be more descriptive of what is does, let me know if you are fine with me changing it.

Yes I have used git actions before, but internally at work we use gitlab and it's runners which I know better.

**Reply by Geolim4** [reply]
> You beat me the the setDoPosixCheck, I was going to use something on the lines of setForkDetection has it would be more descriptive of what is does, let me know if you are fine with me changing it.

Feel free to improve it as I'm unable to test the extension locally atm :)

**Reply by srjlewis** [reply]
I will do the changes a bit later when back near my computer, I just want to finalise the Config API before release :)

**Comment by Geolim4** [comment]
I plan to release the 9.2 for both the core and the extension for 10 January, just so you know :P

**Reply by srjlewis** [reply]
Sounds great, I have rolled it into our internal framework at work and rolled it out across out testing and staging environments, for a bit of battle testing, I it looks like it working well

Just something you might what so look at for the v3 driver

does it need to the following check
https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/f1758025db5c2714003d4f575b62a5d71ddeb15c/lib/Phpfastcache/Extensions/Drivers/Couchbasev4/Driver.php#L154

Also a try catch for users that do not have the right permission for a bucket flush
https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/f1758025db5c2714003d4f575b62a5d71ddeb15c/lib/Phpfastcache/Extensions/Drivers/Couchbasev4/Driver.php#L247-L257

**Reply by Geolim4** [reply]
Thanks, just a simple question:

https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/f1758025db5c2714003d4f575b62a5d71ddeb15c/lib/Phpfastcache/Extensions/Drivers/Couchbasev4/Driver.php#L154
By looking this line, how can a document could be in a "error" state ? It just exist or doesn't ?

**Reply by srjlewis** [reply]
when you remove it you get this error on item the dont exist

```
[FAIL] Uncaught exception "\Couchbase\Exception\DecodingFailureException" in "~/Users/stevenlewis/Documents/PHP Storm Checkouts/couchbasev4-extension/vendor/couchbase/couchbase/Couchbase/JsonTranscoder.php" line 98 with message: "unable to decode bytes with JsonTranscoder"
```
triggered from line  https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/93af157d27c8295eb3f1a370cb7b5b8cc9a1a348/lib/Phpfastcache/Extensions/Drivers/Couchbasev4/Driver.php#L155
<img width="1061" alt="Screenshot at Jan 06 19-09-19" src="https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/56001/d984f8d6-d482-49d8-b344-c27662ec4f92">

so the `$document->error()` indicates there is a error

**Reply by srjlewis** [reply]
If you run the tests with it commented out, you will see

**Reply by Geolim4** [reply]
Is it specific to Couchbasev4 ? I haven't noticed this behavior on v3 ?

**Reply by srjlewis** [reply]
It very well could be, it would not surprise me at all. The permission try catch is more aimed at the couchbase enterprise where the user permissions are very grained.

**Reply by Geolim4** [reply]
> It very well could be, it would not surprise me at all. The permission try catch is more aimed at the couchbase enterprise where the user permissions are very grained.

Ohh right, I only have the community running on my virtual machine as I'm too poor to run an enterprise version 😂
Could you provide me some (censored) screenshoots of the fine-grained permissions so I can see how "fine" are the permissions ?

**Reply by Geolim4** [reply]
Seems that the method "error()" does not exists in Couchbasev3:
![image](https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/1332071/ccc8299a-82bf-4cf8-9899-929dbcec1253)

**Reply by Geolim4** [reply]
Obviously, its a Couchbase 4.0.0 method 😂
![image](https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/1332071/676cf3da-dc6d-4e60-9c80-92702a682420)

**Reply by Geolim4** [reply]
This is getting weirder and weirder:
The document class (when dumped) is a `Couchbase\GetResultImpl` object. 
But I have no reference of this class/interface anywhere ?
And this object seems to implement a `error()` method, since no error is thrown, which make me think that this method is either undocumented or purposely kept secret until v4.

**Reply by srjlewis** [reply]
I was just checking it was needed in v3, but if it is not needed. We are good.

Under their license you are able to install the enterprise and use in for development for free

**Reply by srjlewis** [reply]
I don’t think they have hardly updated any of the docs from v3 and have changed tons of the under laying classes without any docs

**Reply by Geolim4** [reply]
That's funny because v3.2.2 says "error()" does not exist (or is kept secret) while it secretly exists:
https://docs.couchbase.com/sdk-api/couchbase-php-client-3.2.2/classes/Couchbase-GetResult.html
![image](https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/1332071/318a758f-c42b-42ae-a164-fbfe94d138e4)
(Couchbase v3, document dump with Reflection)


While on the v4 it is explicitly documented.
https://docs.couchbase.com/sdk-api/couchbase-php-client-4.1.6/classes/Couchbase-GetResult.html
😖

**Reply by Geolim4** [reply]
Since I have NO idea on which version of Couchbase 3 the "error" method has been added I will not risk to add it in Couchbase v3 😐

**Reply by Geolim4** [reply]
I may keep repeating myself, but for an Enterprise like Couchbase, the lack of documenting and developer support is terribly discouraging to use their product 😖

**Reply by srjlewis** [reply]
We have used it from when it was known as membase, so some times it is the devil you know.

but yes it has bad docs is a under statement.

I will send you some screenshots of the permissions screen, if you don’t want to install the enterprise version

**Reply by Geolim4** [reply]
Enterprise version seems to require you to fill a form with personal data and is 30 days-limited, so for now I'm fine for the Community edition. Thanks for the screenshoots.

**Reply by srjlewis** [reply]
Here is a recording of all the permissions

https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/56001/ea8871f7-9f46-4a20-b847-b444a7171fc2

**Reply by Geolim4** [reply]
That's effectively completely different that the permissions I have in community edition indeed:
![image](https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/1332071/dd22b978-ba7e-47fe-8963-88bb48416114)

**Reply by srjlewis** [reply]
oh yes they are, if I set read and write only, the user wont be able to flush even if the `flushEnabled()` returns true, you only find out when you try and run the command with a nasty ServerInternal exception with a message about permissions, hence the try catch.

**Reply by Geolim4** [reply]
That means even the `driverWrite()` and `driverDelete()` could be impacted too.
But tbh I don't think we need to go that far, catching the flush operation is enough IMO, if there's Exceptions, they will be explicit enough for the user to make the needed changes to Couchbase server configuration.

**Reply by srjlewis** [reply]
I competely agree and that is why I did not put any handling in the `driverWrite()` and `driverDelete()` calls.

**Comment by srjlewis** [comment]
I have put in a test thats tests the lockup of the extention, so hopfully if/when Couchbase fixes it, it will flag it up to us.

**Reply by Geolim4** [reply]
Thanks, judging by the date you created the issue on their tracker, they don't look urged to fix it ? :(

**Comment by Geolim4** [comment]
@srjlewis what is the compatibility threshold limit of Couchbase v3 ? PHP 8.2 or PHP 8.3 ?

**Reply by srjlewis** [reply]
it is PHP 8.2 after that the  sorce code from couchbase repos wont compile and needs patching.

**Reply by Geolim4** [reply]
So it is still compatible for PHP 8.2 ?

Because I hesitate to just deprecate Couchbasev3 for the v10 or remove it since the v10 will increase PHP support to 8.2 minimum.

**Reply by srjlewis** [reply]
in the remi's repo he has v3 and v4 extentions all the way back to PHP 7.4, IMHO I would keep Couchbasev3 but move it to a extention in v10, so it can be tested on its own, but we can keep support for both versions.

**Reply by Geolim4** [reply]
I don't plan to move Couchbasev3 to an extension since it will last for one more version (v10). So IMO it does not worth it.

**Reply by srjlewis** [reply]
true, if you are pushing v10 to PHP 8.2 and most repos install Couchbasev4 by default on PHP 8.2 and 8.3, and some repo wont even have Couchbasev3 for PHP 8.2 and 8.3 it could be worth pull v3 from v10, but a deprecation gives users notice.

But having Couchbasev3 in the core will block testing with PHP 8.3, unless it is in a extention.

I am glad it is your call and not my call (me running away and hiding 😂😂)

**Reply by Geolim4** [reply]
> But having Couchbasev3 in the core will block testing with PHP 8.3, unless it is in a extention.

It is already tested actually so it wont change anything :P

https://github.com/PHPSocialNetwork/phpfastcache/actions/runs/7434678978
https://app.travis-ci.com/github/PHPSocialNetwork/phpfastcache/builds/268172738

> I am glad it is your call and not my call (me running away and hiding 😂😂)

🤣🤣

**Reply by srjlewis** [reply]
ok

travis does not have PHP 8.3 and github action failed to install the couchbase exrtention.
sorry, but testing failed

<img width="1587" alt="Screenshot at Jan 07 00-38-12" src="https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/56001/e3f96d75-3e4c-49cf-80e8-1869989f5eda">
<img width="1020" alt="Screenshot at Jan 07 00-37-24" src="https://github.com/PHPSocialNetwork/couchbasev4-extension/assets/56001/ab79af97-d694-4776-b24c-fa0de6e9c4bc">

**Reply by srjlewis** [reply]
But if you are glad to let the Couchbasev3 test to skip, the rest are fine.

**Reply by Geolim4** [reply]
Its not surprising since I forced the Couchbase v3.2.2 to be installed to test Couchbasev3:

https://github.com/PHPSocialNetwork/phpfastcache/blob/c8a2c7abe2c9d1d429be726c6ff616068df00ef8/.github/workflows/testsv2.yml#L13

Like you said Couchbasev3 is compatible up to PHP 8.2, so the behavior is expected.
Tests passing on 8.0, 8.1, 8.2 for Couchbasev3  is what matters most for me.

Travis support PHP 8.3 as of Ubuntu 22.04. From what I remember we're still on Ubuntu 18.04.
But I plan to remove Travis tests for Phpfastcache v10 in favor of Github actions only, mainly because its pissing me off to ask for Open Source credits every 4~6 months  😖

So I consider it still stable until the v10 at the very last before being removed at v11.

**Reply by Geolim4** [reply]
> But if you are glad to let the Couchbasev3 test to skip, the rest are fine.

They don't pass on PHP 8.3, its expected so its completely fine for me.

**Reply by srjlewis** [reply]
That’s fine and understandable if you happy with it skipping on PHP 8.3, I have never been keen on travis a lot of people are dropping it.

**Reply by Geolim4** [reply]
> That’s fine and understandable if you happy with it skipping on PHP 8.3, I have never been keen on travis a lot of people are dropping it.

I'm not surprised: they removed the ability to use their free services automatically, you have to periodically get in touch with their support to ask for a resupply of 25k min of OSS credit. For Phpfastcache project this credit is spent between 4 and 6 month, imagine for bigger projects.
Also when Github shipped Actions, it was like an earthquake for the Open Source community. Because it is easier to setup, more reliable (travis builds fail a lot sometime for no reasons...), more UI-friendly and the better for the end: Better integrated to the GitHub environment.

I kept both CI services until now for two reasons: Scalability (if one service fails I can rely on another) and because I was afraid of the youthfulness of GitHub action, but there's clearly no reason for me to keep them both anymore, it simply costs me time and brain energy which are both resources that I don't overflow 😂

**Reply by srjlewis** [reply]
🤣 I don’t think I could have put it better myself.

**Reply by Geolim4** [reply]
And yet, knowing you're a fellow England neighbor, I still struggle to write you my best English without making your eyes bleed 😂

**Reply by srjlewis** [reply]
Hey your english is better then mine 🤣🤣🤣

**Reply by Geolim4** [reply]
I'm touched 👉👈  ☺️

**Comment by Geolim4** [comment]
@srjlewis I made a change that I'd like to bring to your attention:
https://github.com/PHPSocialNetwork/couchbasev4-extension/commit/b8f4a38bb2b61caf6b9f625552726218bbf48810

I will make the same mechanism for the all extensions.

Since I found a way to make it SEMVER-compliant I will ship all the extensions for the v9.2.
But they will be composer-required for v9.2, then removed from dependency as of Phpfastcache v10.

This way the changes will remains completely transparent for the developers.

**Reply by srjlewis** [reply]
You have been busy, I think you do need some sleep

instend of useing this code in the config file

https://github.com/PHPSocialNetwork/couchbasev4-extension/blob/7c149b95090411746d238508948b06d926de4d13/tests/Configs/github-actions.php#L4-L11

it would be more readable and eaiser to understand if it was done like this
``` php
return (new Couchbasev4Config())
    ->setItemDetailedDate(true)
    ->setUsername('test')
    ->setPassword('phpfastcache')
    ->setServers(['127.0.0.1])
    ->setBucketName('phpfastcache')
    ->setScopeName('_default')
    ->setCollectionName('_default');
```

what are your thoghts on using the above code.

**Reply by Geolim4** [reply]
You right, its a code I moved from the ConfigFactory class (which will be removed soon) and I forget to remove the closure wrapper 😖

**Reply by Geolim4** [reply]
Busy night indeed. I moved 4 out of 6 driver to their own extension.
Once you have the automatism, it becomes very intuitive 😂

**Reply by srjlewis** [reply]
lol, as I said you need to get some sleep or things get missed, I had a look at the other extentions, looking good 👍

If you need a hand with any others I don't mind, but by the looks of it you are fine and in the groove 😊

**Reply by Geolim4** [reply]
> lol, as I said you need to get some sleep or things get missed, I had a look at the other extentions, looking good 👍

Just finished, I'm a bit proud and I can now go to sleep :P
https://packagist.org/packages/phpfastcache/

> If you need a hand with any others I don't mind, but by the looks of it you are fine and in the groove 😊

Would be my pleasure, tell me on which extension(s) you feel comfortable to help me with !

**Reply by srjlewis** [reply]
Well it looks like 9.2 is looking good, and you should be proud of it.

After the amount of years I have been coding, I can turn my hand to most things, and have used a lot of different technologies. So I would say if you need a different point view, sense check, some coding help or to run some ideas through I dont mind. As its your project is what you are comfortable with.

**Reply by Geolim4** [reply]
Feedback is always appreciated especially if it can improve performances of Phpfastcache in any ways possible !

**Reply by srjlewis** [reply]
I will have a look around, I guess the release got rescheduled from the 10th 🤣 , I will keep a eye out for any issuess, and ping me if you need anything.

**Reply by Geolim4** [reply]
Thank you, i'd say, feel free to test Arangodb/Mongodb/Couchdb extensions if you have free time ahead :P

**Reply by srjlewis** [reply]
when I have a bit of free time I will spin up some VMs in our dev environment, and see if I can switch KV stores without any of my dev's noticing.

**Reply by Geolim4** [reply]
Today I was on a good mood and I created this gem in a couple of hours x)

(It's still on beta stage for now)
https://github.com/PHPSocialNetwork/ravendb-extension

**Reply by srjlewis** [reply]
😁 I do think the extensions are making it a lot more manageable, because you have testing, code and even the docs for each driver on it’s own. And when you do create a new driver it be developed and released without affecting any other parts. But you already know all of that anyway.

I will see if I can test the new extension a bit quicker for you.
